### PR TITLE
Add `saf attest apply` and `saf attest create`

### PR DESCRIPTION
**Pull request by camdenmoors**

### CODE REVIEW

**Comment by camdenmoors** [comment]
Add documentation for sheet name, and empty spreadsheet

File: src/commands/attest/apply.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+      if (Object.entries(executions).length > 1) {
+        fs.mkdirSync(flags.output)
+      }
+
+      if (Object.keys(executions).length === 0) {
+        throw new Error('Please provide at least one HDF file')
+      }
+
+      Object.entries(executions).forEach(([originalFilename, execution]) => {
```
[the remark this answers is not in the document]
Comment (Amndeep7) [reply src/commands/attest/apply.ts:63]:
```suggestion
      for(const [originalFilename, execution] of Object.entries(executions)) {
```

File: src/commands/attest/create.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      const attestations: Attestation[] = []
+
+      if (flags.input) {
+        const evaluation = JSON.parse(fs.readFileSync(flags.input, 'utf8')) as ExecJSON.Execution
+        const search = new AccurateSearch()
+        const controls: Record<string, ExecJSON.Control> = {}
+        for (const profile of evaluation.profiles) {
+          for (const control of profile.controls) {
+            controls[control.id] = control
+            search.addText(control.id, control.id + ': ' + (control.title || '') + ' ' + control.desc || '')
```
Comment (Amndeep7) [inline src/commands/attest/create.ts:70]: shouldn't contro.desc || '' also be in parentheses?

File: src/commands/attest/create.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        while (true) {
+          const input = prompt("Enter a control ID, search for a control, or enter 'q' to exit: ")
+          if (input.trim().toLowerCase() === 'q') {
+            break
+          } else if (input in controls) {
+            const control = controls[input]
+            const attestation = this.promptForAttestation(control.id)
+            attestations.push(attestation)
```
Comment (Amndeep7) [inline src/commands/attest/create.ts:81]: can't we make this a one-liner?

File: src/commands/attest/create.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+          } else {
+            const ids = search.search(input).slice(0, MAX_SEARCH_RESULTS)
+            ids.forEach((id: string) => {
```
Comment (Amndeep7) [inline src/commands/attest/create.ts:84]:
```suggestion
            for(const id of ids as string[]) {
```

File: src/commands/attest/create.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+      switch (flags.format) {
+      case 'json': {
+        fs.writeFileSync(flags.output, JSON.stringify(attestations, null, 2))
+        break
+      }
+
+      case 'xlsx': {
+        XlsxPopulate.fromDataAsync(dataURLtoU8Array(files.AttestationTemplate.data)).then((workBook: any) => {
+          const sheet = workBook.sheet(0)
```
Comment (Amndeep7) [inline src/commands/attest/create.ts:109]: probably good to leave a comment that sheet 0 is intended to be that explicitly named attestations sheet

File: src/commands/attest/create.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+          let currentRow = 2
+          attestations.forEach(attestation => {
```
Comment (Amndeep7) [inline src/commands/attest/create.ts:111]:
```suggestion
          for(const attestation of attestations) {
```

File: test/commands/attest/apply.test.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import {expect, test} from '@oclif/test'
+import * as tmp from 'tmp'
+import path from 'path'
+import fs from 'fs'
+import {ExecJSON} from 'inspecjs'
+import _ from 'lodash'
+
+export function omitHDFChangingFields(
```
Comment (Amndeep7) [inline test/commands/attest/apply.test.ts:8]: we should probably put this (and similar functions that are currently just being duplicated like omitVersions) into a utils file like how we do in hdfconverters

Comment (Amndeep7) [reply test/commands/attest/apply.test.ts:8]: or maybe replace the other files' omitversions implementation with this one after it's moved into a utils file

**Review by Amndeep7 — CHANGES_REQUESTED** [review]

File: src/commands/attest/create.ts (hunk 2 above)
Comment (camdenmoors) [reply src/commands/attest/create.ts:70]: Because it's at the end of the function call it works without this

Comment (Amndeep7) [reply src/commands/attest/create.ts:70]: I disagree: 

```
> a
{ a: 'a', b: 'b' }
> a.a + ": " + (a.c || '') + " " + a.d || ''
'a:  undefined'
```

That control.desc could undefined or null and consequently would show up like that in the string.  This is cause logical OR is way below addition in terms of operator precedence.

Tangent: that generated string should probs also be a template string anyways

File: src/commands/attest/create.ts [hunk 8]
```diff
@@ -67,7 +67,7 @@ export default class CreateAttestations extends Command {
         for (const profile of evaluation.profiles) {
           for (const control of profile.controls) {
             controls[control.id] = control
-            search.addText(control.id, control.id + ': ' + (control.title || '') + ' ' + control.desc || '')
+            search.addText(control.id, `${control.id} ${control.title} ${control.desc}`)
```
Comment (Amndeep7) [inline src/commands/attest/create.ts:70]: not a match with the previous behavior (we're losing the colon there) and also doesn't address the issue of what happens if one of those attributes is missing
```suggestion
            search.addText(control.id, `${control.id}: ${control.title || ''} ${control.desc || ''}`)
```

**Review by Amndeep7 — CHANGES_REQUESTED** [review]

**Comment by blake-burton** [comment]
@Amndeep7 I was able to get saf attest apply to work with a json file, but not with a xlsx file. i put the same control in the xlsx file and in json file using the format from the test file from this PR. when i run 

```
docker run -v $(pwd):/share mitre/saf:latest attest apply -i in.hdf attestation.json -o out.hdf 
docker run -v $(pwd):/share mitre/saf:latest attest apply -i in.hdf attestation.xlsx -o out.hdf 
```
only the json input produced the attestation_data in the output

 while looking at the tests in this PR, i did not see attestation_data in the xlsx test output. doesnt look like any of the controls in the xlsx are in the sample output, which makes since why there is no attestation_data. is this intended?

**Comment by Amndeep7** [comment]
Probably not haha.  @camdenmoors and @Rlin232 can y'all take a look at this issue and also the thing that Ryan found with the tests?

**Comment by camdenmoors** [comment]
> @camdenmoors and @Rlin232 can y'all take a look at this issue and also the thing that Ryan found with the tests?

The issue seems to be that the spreadsheet used in tests actually has controls for a RHEL7 profile instead of the oracle DB in triple_overlay_example. Using the correct results does apply attestations on my end:
attested

@blake-burton were you using the sample data to test this out? that may be the issue

**Comment by blake-burton** [comment]
@camdenmoors, I was able to get the attestations to work when using the examples, but still didnt work for the examples i had. after some poking around turns out it was the tab name :(. when originally trying this out i used the example xlsx from here https://github.com/mitre/inspec-reporter-json-hdf/tree/main/samples, and the tab name is not "attestations".  is there other documentation to layout the format of the xlsx file?

**Pull request merged**